internal/pool/pool.go

```diff
@@ -358,25 +397,29 @@ func (p *Pool[PT, T]) putItem(ctx context.Context, item PT) (finalErr error) {
 	select {
 	case <-p.done:
 		return xerrors.WithStackTrace(errClosedPool)
+	case <-ctx.Done():
+		return xerrors.WithStackTrace(ctx.Err())
 	default:
-		if !item.IsAlive() {
+		p.stats.InUse().Dec()
```

Comment:
manage counters must be outside from putItem call

Reply:
In this approach Idle() counter is showing amount of available items in queue channel, and because of this they are changed right after channel operation in getItem/putItem. Since getItem/putItem abstract queue interaction for try() and may discard item (instead of returning it (in get) or putting back to queue (in put)), managing Idle() counter outside of put/get will result in wrong value (for example when discarded items are still counted as idle).

There's not harm in moving InUse() counter outside. (Did it in last commit)

To sum up, I think It is fine if in some cases Idle+InUse < Limit, since imo counters represent stats about existing items. If item is discarded it should not be represented in stats.